contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -1588,6 +1590,245 @@ create_xp_instance_regread_in_master_dbo_internal(PG_FUNCTION_ARGS)
 	PG_RETURN_INT32(0);
 }
 
+Datum
+sp_execute_postgresql(PG_FUNCTION_ARGS)
+{
+	List	   *parsetree_list;
+	ListCell   *parsetree_item;
+	char	   *extensionStmt;
+	Node	   *stmt;
+	size_t		len;
+	const char *saved_dialect = GetConfigOption("babelfishpg_tsql.sql_dialect", true, true);
+	Oid			current_user_id = GetUserId();
+	const char *saved_buffer = pstrdup(GetConfigOption("search_path", true, true));
+	const char *new_buffer = "public, %s, \"$user\", sys, pg_catalog";
+	
+	PG_TRY();
+	{	
+		set_config_option("babelfishpg_tsql.sql_dialect", "postgres",
+						GUC_CONTEXT_CONFIG,
+						PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
+		
+		extensionStmt = PG_ARGISNULL(0) ? NULL : TextDatumGetCString(PG_GETARG_TEXT_PP(0));
+
+		if (extensionStmt == NULL)
+				ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+								errmsg("Statement cannot be NULL.")));
+
+		/* Remove trailing whitespaces */
+			len = strlen(extensionStmt);
+			while (isspace(extensionStmt[len - 1]))
+				extensionStmt[--len] = 0;
+
+			len = strlen(extensionStmt);
+
+		/* check if input statement is empty after removing trailing spaces */
+			if (len == 0)
+				ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+											errmsg("Statement cannot be NULL.")));
+
+		parsetree_list = raw_parser(extensionStmt, RAW_PARSE_DEFAULT);
```

Comment:
Instead of calling `raw_parser` and `ProcessUtility`, did we consider using `SPI`, e.g., https://www.postgresql.org/docs/current/spi-spi-execute.html?

Reply:
Yes, we did consider this as one of the approaches but since we are only supporting utility statements for now, the raw_parser approach seems to suit more.